input/drivers/android_input.c

```diff
@@ -869,7 +871,7 @@ static void handle_hotplug(android_input_data_t *android_data,
    if (!back_mapped && settings->input.back_as_menu_toggle_enable)
       settings->input.autoconf_binds[*port][RARCH_MENU_TOGGLE].joykey = AKEYCODE_BACK;
 
-   android_data->pad_states[android_data->pads_connected].id = id;
+   android_data->pad_states[*port].id = id;
```

Comment:
I'm not sure the whole thing is a bit fragile. That's why all the hacks are enclosed and isolated from the normal input flow.
I'll try with the current code and two shield controllers but I tested back then and it was working fine (setting port to 0 will make it overwrite everything anyway)